src/libraries/System.Private.CoreLib/src/System/IO/File.cs

```diff
@@ -251,27 +251,33 @@ public static void WriteAllText(string path, string? contents, Encoding encoding
 
         public static byte[] ReadAllBytes(string path)
         {
-            // bufferSize == 1 used to avoid unnecessary buffer in FileStream
-            using (FileStream fs = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 1, FileOptions.SequentialScan))
+            // SequentialScan is a perf hint that requires extra sys-call on non-Windows OSes.
+            FileOptions options = OperatingSystem.IsWindows() ? FileOptions.SequentialScan : FileOptions.None;
+            using (SafeFileHandle sfh = OpenHandle(path, FileMode.Open, FileAccess.Read, FileShare.Read, options))
             {
                 long fileLength = 0;
-                if (fs.CanSeek && (fileLength = fs.Length) > int.MaxValue)
+                if (sfh.CanSeek && (fileLength = RandomAccess.GetFileLength(sfh)) > Array.MaxLength)
                 {
                     throw new IOException(SR.IO_FileTooLong2GB);
                 }
+
+#if DEBUG
+                fileLength = 0; // improve the test coverage for ReadAllBytesUnknownLength
+#endif
```

Comment:
> Can we instead just ensure we have tests that read files that return 0 for the length?

We have some tests, and you are the person who authored most of them: https://github.com/dotnet/corefx/pull/28388

The problem is that they don't guarantee that all possible code paths are going to be executed (using stack-allocated array, renting an array from the pool, returning it and renting a larger array etc) as they deal with a virtual file system that seems to be exposing files only for reading.

We could use pipes (#58434 introduced some tests for that ), but it would require more work and handling all edge cases.

I know that setting `fileLength` to `0` with `#if DEBUG` is ugly, but it gives us what we need with a very little effort.

Reply:
> but it gives us what we need with a very little effort.

Does it?  It means, for example, that Debug.Asserts in the normal non-zero code paths are now rendered useless.

Reply:
> It means, for example, that Debug.Asserts in the normal non-zero code paths are now rendered useless.

Which asserts do you mean? I can't see any in the code path for regular files that report length properly:

https://github.com/dotnet/runtime/blob/25763904ee0f3dac6b9769b4c3479d4aee2f438d/src/libraries/System.Private.CoreLib/src/System/IO/File.cs#L548-L567

https://github.com/dotnet/runtime/blob/25763904ee0f3dac6b9769b4c3479d4aee2f438d/src/libraries/System.Private.CoreLib/src/System/IO/File.cs#L275-L289

Reply:
> Which asserts do you mean?

Any on any code paths used from here all the way down through the runtime.

Reply:
> Any on any code paths used from here all the way down through the runtime.

These paths (namely `RandomAccess.ReadAtOffsetAsync` and `RandomAccess.ReadAtOffset`) have an excellent test code coverage (directly as `RandomAccess` tests and indirectly via `FileStream` tests) so I am not worried about it.

I do understand that it's not a clear win and a a trade off.